### ISSUE TITLE
diagram images are never generated into _site and are not copied automatically, so they get lost

### ISSUE THREAD

**Issue by a4z** [asciidoctor/jekyll-asciidoc#156]
it is impossible to generate, or get auto copied, diagram images into _site
you need to generate the into somewhere else, 
and run jekyll build again so that jekyll copies them into _site.

```
asciidoctor:
  base_dir: :source
  safe: unsafe
  attributes:
    imagesdir: /images
    imagesoutdir: foo/images
```
generates images into foo/images (forget about that imgesdir is not correct now, it's not relevant in this context)
using this
```
asciidoctor:
  base_dir: :source
  safe: unsafe
  attributes:
    imagesdir: /images
    imagesoutdir: _site/images
```
there will be no images

This requires to run `bundle exec jekyll s` 2 times so that the images land in the destination directory
Very error prone, forget live reload or even auto re-generate. And running this on a build server which I can not re configure to run twice loses the images

I am unsure if this is a problem of this plugin, or jekyll in general and if jekyll requires a fix so that this would work

**Comment by mojavelinux** [comment]
There are likely two problems you are facing.

First, make sure you are using Asciidoctor Diagram 1.5.4, which automatically configures the output directory to the location where other files are being generated. (Essentially, it auto-configures imagesoutdir so you don't have to manage it).

Second, Jekyll has a long-standing bug which deletes generated files in the middle of the workflow. There are two ways to bypass it. The simplest is to configure "keep files", which are files or directories in the _site directory that it will not touch. Here, we tell it to ignore the images directory.

_config.yml
```yaml
keep_files:
  - images
```

The other workaround is to create a plugin that disables the cleaner altogether.

_plugins/jekyll-ext.rb
```ruby
class Jekyll::Cleaner
  def cleanup!; end
end
```

The first approach is the simplest.

I went ahead and put an end-to-end tutorial together to show how to get Jekyll and Asciidoctor Diagram working together. Likely, however, the only part you were missing was the keep_files configuration.

See https://gist.github.com/mojavelinux/968623c493190dd61c059c2d85f9bdc3

**Comment by mojavelinux** [comment]
I'll update the README to include these points and link to that gist.

**Comment by mojavelinux** [comment]
> Very error prone, forget live reload or even auto re-generate.

To be honest, it's really the Jekyll cleaner that is messing everything up. I really don't know what is going on there, but they need to rethink why it is running in the middle of a build.

**Comment by mojavelinux** [comment]
README updated!

Let me know if those instructions work for you. I'm happy to update if there are still details missing.

**Comment by mojavelinux** [comment]
> diagram images are never generated into _site and are not copied automatically, so they get lost

To clarify, the diagram images are generated into _site, and thus are copied automatically. But they do get lost by default. That's because of the Jekyll cleaner.

**Comment by a4z** [comment]
Wow, thank you so much Dan for spending your weekend time on this issue!

I also think that the actual problem is that Jekyll does not care too much about generated files, and some other things.

I will use your guide when I have again access to the documents on Monday and report the result.
What I already can say it that the updated README looks great and the kick start example looks excellent, this makes everything so much clearer for me, thanks!

**Comment by mojavelinux** [comment]
>  the actual problem is that Jekyll does not care too much about generated files, and some other things.

Well said.

**Comment by mojavelinux** [comment]
> What I already can say it that the updated README looks great and the kick start example looks excellent, this makes everything so much clearer for me, thanks!

I'm glad to hear it! Your feedback really helps me know what to focus on in the documentation, so keep it coming!

**Comment by a4z** [comment]
It works as in the it stands now in the README, 
The new text and the gist template makes it super easy to follow.
Thanks a lot four your help!

**Comment by mojavelinux** [comment]
It makes me so pleased to hear that! If you get stuck on anything else, don't hesitate to raise an issue and I'll do my best to improve the docs (or code, if necessary). Good luck!

**Comment by MariuszCwikla** [comment]
All is good, but it does not work with `baseurl`. I tried to use Gitlab Pages to deploy blog using default template (see https://docs.gitlab.com/ee/user/project/pages/getting_started/pages_bundled_template.html and https://gitlab.com/pages/jekyll)
Gitlab Pages requires you to use `baseurl: "/something"`.
But then links to images get broken. Changing to `imagesdir: /something/images` does not help.
I tried multiple combinations - files are either removed or not copied or jekyll runs into infinite loop of copying files.

However I solved this issue by having 2 config files:

* `_config.yml` - used by gitlab pipelines. I also need to copy some files manually.
* `_config-dev.yml` - I use this for local development

1. `_config-dev.yml` - defined exactly as per your tips and **empty** `baseurl`

```
baseurl: ""     #MUST BE EMPTY
keep_files:
  - images
asciidoctor:
  base_dir: :docdir
  safe: unsafe
  attributes:
    imagesdir: /images
```

Then locally I use `jekyll serve --config _config-dev.yml`

2. Gitlab pipeline uses regular `_config.yml` file:

```
baseurl: "/blog"
keep_files:
  - images
asciidoctor:
  base_dir: :docdir
  safe: unsafe
  attributes:
    imagesdir: /blog/images
    imagesoutdir: images

```

and need to copy generated images manually in `.gitlab-ci.yml`:

```
  - bundle exec jekyll build -d public  
  - cp -r _posts/images/ public         #HERE
```
